airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/components.py

```diff
@@ -0,0 +1,362 @@
+from airbyte_cdk.sources.streams.http import HttpClient
+from airbyte_cdk.sources.streams.http.exceptions import DefaultBackoffException, RequestBodyException, UserDefinedBackoffException
+from airbyte_cdk.sources.streams.http.http import BODY_REQUEST_METHODS
+from isodate import Duration, parse_duration
+
+from .utils import ANALYTICS_FIELDS_V2, FIELDS_CHUNK_SIZE, transform_data
+
+
+class SafeHttpClient(HttpClient):
+    def _create_prepared_request(
+        self,
+        http_method: str,
+        url: str,
+        dedupe_query_params: bool = False,
+        headers: Optional[Mapping[str, str]] = None,
+        params: Optional[Mapping[str, str]] = None,
+        json: Optional[Mapping[str, Any]] = None,
+        data: Optional[Union[str, Mapping[str, Any]]] = None,
+    ) -> requests.PreparedRequest:
+        if dedupe_query_params:
+            query_params = self._dedupe_query_params(url, params)
+        else:
+            query_params = params or {}
+        query_params = urlencode(query_params, safe="():,%")
+        args = {"method": http_method, "url": url, "headers": headers, "params": query_params}
+        if http_method.upper() in BODY_REQUEST_METHODS:
+            if json and data:
+                raise RequestBodyException(
+                    "At the same time only one of the 'request_body_data' and 'request_body_json' functions can return data"
+                )
+            elif json:
+                args["json"] = json
+            elif data:
+                args["data"] = data
+        prepared_request: requests.PreparedRequest = self._session.prepare_request(requests.Request(**args))
+
+        return prepared_request
+
+
+@dataclass
+class SafeEncodeHttpRequester(HttpRequester):
+    """
+    This custom component safely validates query parameters, ignoring the symbols ():,% for UTF-8 encoding.
+
+    Attributes:
+        request_body_json: Optional JSON body for the request.
+        request_headers: Optional headers for the request.
+        request_parameters: Optional parameters for the request.
+        request_body_data: Optional data body for the request.
+    """
+
+    request_body_json: Optional[RequestInput] = None
+    request_headers: Optional[RequestInput] = None
+    request_parameters: Optional[RequestInput] = None
+    request_body_data: Optional[RequestInput] = None
+
+    def __post_init__(self, parameters: Mapping[str, Any]) -> None:
+        self.request_options_provider = InterpolatedRequestOptionsProvider(
+            request_body_data=self.request_body_data,
+            request_body_json=self.request_body_json,
+            request_headers=self.request_headers,
+            request_parameters=self.request_parameters,
+            config=self.config,
+            parameters=parameters or {},
+        )
+        super().__post_init__(parameters)
+
+        if self.error_handler is not None and hasattr(self.error_handler, "backoff_strategies"):
```

Comment:
I'm sure there is a reason why we check for `is not None`, but can we:

```python
if self.error_handler and hasattr(self.error_handler, "backoff_strategies"):
    ....
```

Reply:
1. Save consistency with parent implementation
2. With `if self.error_handler` you call __bool__ or __len__ methods if `self.error_handler` isn't `None`, but when `if self.error_handler is not None` you don't call anything and just use `is` operator behavior.